odoo/fields.py

```diff
@@ -2423,32 +2423,37 @@ def _setup_attrs(self, model_class, name):
                 new_values = [kv[0] for kv in selection_add if kv[0] not in values]
                 for key in new_values:
                     ondelete.setdefault(key, 'set null')
-                if self.required and new_values and 'set null' in ondelete.values():
-                    raise ValueError(
-                        "%r: required selection fields must define an ondelete policy that "
-                        "implements the proper cleanup of the corresponding records upon "
-                        "module uninstallation. Please use one or more of the following "
-                        "policies: 'set default' (if the field has a default defined), 'cascade', "
-                        "or a single-argument callable where the argument is the recordset "
-                        "containing the specified option." % self
-                    )
 
                 # check ondelete values
                 for key, val in ondelete.items():
-                    if callable(val) or val in ('set null', 'cascade'):
+                    if callable(val) or val == 'cascade':
                         continue
-                    if val == 'set default':
+                    elif val == 'set null':
+                        assert not self.required, (
+                            "%r: required selection fields must define an ondelete policy that "
+                            "implements the proper cleanup of the corresponding records upon "
+                            "module uninstallation. Please use one or more of the following "
+                            "policies: 'set default' (if the field has a default defined), "
+                            "'cascade', or a single-argument callable where the argument is the "
+                            "recordset containing the specified option."
+                        ) % self
+                    elif val == 'set default':
                         assert self.default is not None, (
                             "%r: ondelete policy of type 'set default' is invalid for this field "
                             "as it does not define a default! Either define one in the base "
                             "field, or change the chosen ondelete policy" % self
                         )
-                        continue
-                    raise ValueError(
-                        "%r: ondelete policy %r for selection value %r is not a valid ondelete "
-                        "policy, please choose one of 'set null', 'set default', 'cascade' or "
-                        "a callable" % (self, val, key)
-                    )
+                    elif val.startswith('set '):
+                        assert val[4:] in dict(self.selection), (
```

Comment:
At that point, `self.selection` is actually not valid
```suggestion
                        assert val[4:] in values, (
```